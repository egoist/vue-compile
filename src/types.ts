export interface ScriptCompilerContext {
  filename: string
  babelrc?: boolean
  modern?: boolean
}
