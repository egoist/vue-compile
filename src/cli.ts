#!/usr/bin/env node
import chalk from 'chalk'

if (parseInt(process.versions.node, 10) < 8) {
  console.error(
    chalk.red('The "vue-compile" module requires Node.js 8 or above!')
  )
  console.error(chalk.dim(`Current version: ${process.versions.node}`))
  process.exit(1)
}

async function main(): Promise<void> {
  const { cac } = await import('cac')
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { version } = require('../package.json')

  const cli = cac('vue-compile')

  cli
    .command('[input]', 'Normalize input file or directory', {
      ignoreOptionDefaultValue: true
    })
    .usage('[input] [options]')
    .action(async (input: string, flags: any) => {
      const options = {
        input,
        ...flags
      }

      if (!options.input) {
        delete options.input
      }

      if (options.debug === true) {
        process.env.DEBUG = 'vue-compile:*'
      } else if (typeof options.debug === 'string') {
        process.env.DEBUG = `vue-compile:${options.debug}`
      }

      const { createCompiler } = await import(".")

      if (!options.input) {
        return cli.outputHelp()
      }

      const compiler = createCompiler(options)

      compiler.on('normalized', async (input: string, output: string) => {
        if (!compiler.options.debug) {
           const { humanlizePath } = await import('./utils')

          console.log(
            `${chalk.magenta(humanlizePath(input))} ${chalk.dim(
              '->'
            )} ${chalk.green(humanlizePath(output))}`
          )
        }
      })

      return compiler.normalize().catch(handleError)
    })
    .option('-o, --output <file|directory>', 'Output path')
    .option(
      '-i, --include <glob>',
      'A glob pattern to include from input directory'
    )
    .option(
      '-e, --exclude <glob>',
      'A glob pattern to exclude from input directory'
    )
    .option('--no-babelrc', 'Disable .babelrc file')
    .option(
      '--modern',
      'Only supports browsers that support <script type="module">'
    )

  cli.option('--debug', 'Show debug logs')

  cli.version(version)
  cli.help()

  cli.parse()
}

function handleError(error: Error): void {
  console.error(error.stack)
  process.exit(1)
}

main().catch(handleError)
