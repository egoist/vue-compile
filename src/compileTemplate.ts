import { SFCBlock } from '@vue/component-compiler-utils'

export const compileTemplate = <T = SFCBlock | null>(template: T): T => template
