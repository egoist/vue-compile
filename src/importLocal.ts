import resolveFrom from 'resolve-from'

export const importLocal = (dir: string, name: string, fallback?: string): any => {
  const found =
    resolveFrom.silent(dir, name) ||
    (fallback && resolveFrom.silent(dir, fallback))

  if (!found) {
    throw new Error(
      `You need to install "${name}"${
        fallback ? ` or "${fallback}"` : ''
      } in current directory!`
    )
  }

  return require(found)
}
