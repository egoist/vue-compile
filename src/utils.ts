import path from 'path'

export const humanlizePath = (p: string): string => path.relative(process.cwd(), p)

export const notSupportedLang = (lang: string, tag: string): string => {
  return `"${lang}" is not supported for <${tag}> tag currently, wanna contribute this feature?`
}

function escapeRe(str: string): string {
  return str.replace(/[-[\]/{}()*+?.\\^$|]/g, '\\$&')
}

export const replaceContants = (content: string, constants?: {[k:string]:any}): string => {
  if (!constants) return content

  const RE = new RegExp(
    `\\b(${Object.keys(constants)
      .map(escapeRe)
      .join('|')})\\b`,
    'g'
  )
  content = content.replace(RE, (_, p1) => {
    return constants[p1]
  })

  return content
}

export const cssExtensionsRe = /\.(css|s[ac]ss|styl(us)?)$/
