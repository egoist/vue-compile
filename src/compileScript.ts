import { SFCBlock } from '@vue/component-compiler-utils'
import { notSupportedLang } from './utils'
import { ScriptCompilerContext } from './types'

export const compileScript = async (
  script: SFCBlock | null,
  ctx: ScriptCompilerContext
): Promise<SFCBlock | null> => {
  if (!script) return script

  const code = script.content.replace(/^\/\/$/gm, '')

  if (script.lang === 'ts' || script.lang === 'typescript') {
    script.content = await import('./script-compilers/ts').then(
      async ({ compile }) => compile(code, ctx)
    )
  } else if (
    !script.lang ||
    script.lang === 'esnext' ||
    script.lang === 'babel'
  ) {
    script.content = await import('./script-compilers/babel').then(
      async ({ compile }) => compile(code, ctx)
    )
  } else {
    throw new Error(notSupportedLang(script.lang, 'script'))
  }

  return script
}
