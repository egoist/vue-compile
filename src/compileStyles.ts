import { SFCBlock } from '@vue/component-compiler-utils'
import { notSupportedLang } from './utils'

export const compileStyles = async (
  styles: SFCBlock[],
  { filename }: { filename: string }
): Promise<SFCBlock[]> => {
  return Promise.all(
    styles.map(async style => {
      // Do not handle "src" import
      // Until we figure out how to handle it
      if (style.src) return style

      const { content } = style

      if (style.lang === 'stylus') {
        style.content = await require('./style-compilers/stylus').compile(content, {
          filename
        })
      } else if (!style.lang || style.lang === 'postcss') {
        style.content = await require('./style-compilers/postcss').compile(content, {
          filename
        })
      } else if (style.lang === 'scss' || style.lang === 'sass') {
        style.content = await require('./style-compilers/sass').compile(content, {
          filename,
          indentedSyntax: style.lang === 'sass'
        })
      } else if (style.lang) {
        throw new Error(notSupportedLang(style.lang, 'style'))
      }

      return style
    })
  )
}
