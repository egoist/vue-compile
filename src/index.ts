import path from 'path'
import { EventEmitter } from 'events'
import fs from 'fs-extra'
import JoyCon from 'joycon'
import isBinaryPath from 'is-binary-path'
import createDebug from 'debug'
import { replaceContants, cssExtensionsRe } from './utils'

const debug = createDebug('vue-compile:cli')

interface OptionContants {
  [k: string]: string | boolean | number
}

interface InputOptions {
  config?: boolean | string
  input: string
  output: string
  constants?: OptionContants
  modern?: boolean
  babelrc?: boolean
  include?: string[]
  exclude?: string[]
  debug?: boolean
}

interface NormalizedOptions {
  input: string
  output: string
  constants?: OptionContants
  modern?: boolean
  babelrc?: boolean
  include?: string[]
  exclude?: string[]
  debug?: boolean
}

class VueCompile extends EventEmitter {
  options: NormalizedOptions

  isInputFile?: boolean

  constructor(options: InputOptions) {
    super()

    if (options.config !== false) {
      const joycon = new JoyCon({
        files: [
          typeof options.config === 'string'
            ? options.config
            : 'vue-compile.config.js'
        ],
        stopDir: path.dirname(process.cwd())
      })
      const { data: config, path: configPath } = joycon.loadSync()

      if (configPath) {
        debug(`Using config file: ${configPath}`)
      }

      options = { ...options, ...config}
    }

    this.options = this.normalizeOptions(options)
    this.isInputFile = fs.statSync(this.options.input).isFile()
  }

  normalizeOptions(options: InputOptions): NormalizedOptions {
    return { ...options, input: path.resolve(options.input),
      output: path.resolve(options.output)}
  }

  async normalize(): Promise<void> {
    if (!this.options.input) {
      return
    }

    if (this.isInputFile) {
      if (!this.options.output) {
        throw new Error('You must specify the path to output file.')
      }

      await this.normalizeFile(this.options.input, this.options.output)
    } else {
      if (!this.options.output) {
        throw new Error('You must specify the path to output directory.')
      }

      await this.normalizeDir(this.options.input, this.options.output)
    }
  }

  async normalizeFile(input: string, outFile: string): Promise<void> {
    if (isBinaryPath(input)) {
      const buffer = await fs.readFile(input)
      return this.writeBinary(buffer, {
        filename: input,
        outFile
      })
    }

    let source = await fs.readFile(input, 'utf8')
    source = replaceContants(source, this.options.constants)

    const ctx = {
      filename: input,
      outFile,
      modern: this.options.modern,
      babelrc: this.options.babelrc
    }

    if (!input.endsWith('.vue')) {
      return this.writeText(source, ctx)
    }

    const sfcDescriptor = await import('@vue/component-compiler-utils').then(
      ({ parse }) => {
        return parse({
          compiler: require('vue-template-compiler'),
          source,
          filename: input,
          needMap: false
        })
      }
    )

    const script = await import('./compileScript').then(async ({ compileScript }) =>
      compileScript(sfcDescriptor.script, ctx)
    )
    const template = await import('./compileTemplate').then(
      async ({ compileTemplate }) => compileTemplate(sfcDescriptor.template)
    )
    const styles = await import('./compileStyles').then(async ({ compileStyles }) =>
      compileStyles(sfcDescriptor.styles, ctx)
    )

    await import('./writeSFC').then(async ({writeSFC}) => writeSFC(
      {
        script,
        styles,
        template,
        customBlocks: sfcDescriptor.customBlocks
      },
      outFile
    ))

    this.emit('normalized', input, outFile)
  }

  async normalizeDir(input: string, outDir: string): Promise<void> {
    const include = [...(this.options.include || [])]
    const exclude = [...(this.options.exclude || [])]
    const { default: glob } = await import('fast-glob')
    const files = await glob(
      include.length > 0 ? include : ['**/*'],
      {
        cwd: input,
        ignore: ['**/node_modules/**'].concat(exclude)
      }
    )
    await Promise.all(
      files.map(async (file: string) => {
        return this.normalizeFile(
          path.join(input, file),
          path.join(outDir, file)
        )
      })
    )
  }

  async writeText(
    source: string,
    {
      filename,
      babelrc,
      modern,
      outFile
    }: {
      filename: string
      outFile: string
      babelrc?: boolean
      modern?: boolean
    }
  ): Promise<void> {
    let output

    if (filename.endsWith('.js')) {
      output = await import('./script-compilers/babel').then(async ({ compile }) => {
        return compile(source, {
          filename,
          babelrc,
          modern
        })
      })
    } else if (filename.endsWith('.ts')) {
      output = await import('./script-compilers/ts').then(async ({ compile }) => {
        return compile(source, {
          filename,
          babelrc,
          modern
        })
      })
      outFile = outFile.replace(/\.ts$/, '.js')
    } else if (filename.endsWith('.css')) {
      output = await import('./style-compilers/postcss').then(async ({ compile }) => {
        return compile(source, { filename })
      })
    } else if (/\.s[ac]ss$/.test(filename)) {
      const basename = path.basename(filename)
      if (basename.startsWith('_')) {
        // Ignore sass partial files
        return
      }

      output = await import('./style-compilers/sass').then(async ({ compile }) => {
        return compile(source, {
          filename,
          indentedSyntax: filename.endsWith(".sass")
        })
      })
    } else if (/\.styl(us)?/.test(filename)) {
      output = await import('./style-compilers/stylus').then(async ({ compile }) => {
        return compile(source, {
          filename
        })
      })
    } else {
      output = source
    }

    outFile = outFile.replace(cssExtensionsRe, '.css')

    await fs.outputFile(outFile, output, 'utf8')

    this.emit('normalized', filename, outFile)
  }

  async writeBinary(
    source: Buffer,
    { filename, outFile }: { filename: string; outFile: string }
  ): Promise<void> {
    await fs.outputFile(outFile, source, 'utf8')

    this.emit('normalized', filename, outFile)
  }
}

export const createCompiler = (opts: InputOptions): VueCompile => new VueCompile(opts)
