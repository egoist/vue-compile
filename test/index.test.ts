import path from 'path'
import fs from 'fs-extra'
import { createCompiler } from '../src'

const fixture = (...args: string[]): string => path.resolve(__dirname, 'fixture', ...args)
const tmp = (name: string): string => fixture('output', name)

test('custom blocks', async () => {
  const outDir = tmp('custom-blocks')
  const compiler = createCompiler({
    input: fixture('custom-blocks'),
    output: outDir
  })
  await compiler.normalize()
  const content = await fs.readFile(path.join(outDir, 'foo.vue'), 'utf8')
  expect(content).toMatchSnapshot()
})

