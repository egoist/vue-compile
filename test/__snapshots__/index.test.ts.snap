// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom blocks 1`] = `
"<template>  
  <div></div>
</template>

<script>
export default {};
</script>

<foo>this is a custom block</foo>"
`;
